### ISSUE TITLE
resolved DNSSEC validation can be bypassed by MITM

### ISSUE THREAD

**Issue by AgentOak** [systemd/systemd#25676]
### systemd version the issue has been seen with

251.8-1-manjaro

### Used distribution

Manjaro

### Linux kernel version used

5.15.81-1-MANJARO

### CPU architectures issue was seen on

x86_64

### Component

systemd-resolved

### Expected behaviour you didn't see

When `DNSSEC=yes` is set and querying a DNSSEC-enabled domain, systemd-resolved should only accept records that are signed correctly.

### Unexpected behaviour you saw

systemd-resolved accepts records of DNSSEC-signed domains even when they have no signature, allowing man-in-the-middles (or the upstream DNS resolver) to manipulate records.

Please note that I have repeatedly tried to report this issue to the systemd-security mailing list. I sent a mail with the details and reproduction steps to the systemd-security address on Oct 6th, Oct 14th and Oct 22nd. I first received a reply on Oct 24th from Lennart Poettering, asking for some basic system information, which I provided on the same day. Then I received no further reply, so I sent the report one more time to systemd-security on Nov 7th. The next day Poettering replied to my previous mail, asking for a debug log, but not directly acknowledging the issue. I sent debug logs on Nov 10th and haven't heard back from anyone since then.

### Steps to reproduce the problem

The issue can be reproduced with an unbound server and a typetransparent local-zone to simulate an attacker modifying select records. `cloudflare.net` is used as an example domain since it is [signed and valid](https://dnssec-analyzer.verisignlabs.com/cloudflare.net). In this example, the DNS server is run on `192.168.1.1` and the systemd client has any address in the same /24 network.

unbound.conf:
```
server:
  # Allow running unbound unprivileged
  username: ""
  use-syslog: no
  chroot: ""
  pidfile: ""

  interface: 0.0.0.0
  port: 10053
  access-control: 192.168.1.1/24 allow

  local-zone: www.cloudflare.net typetransparent
  local-data: "www.cloudflare.net. 3600 IN A 1.2.3.4"

forward-zone:
  name: .
  forward-addr: 1.1.1.1
  forward-addr: 1.0.0.1
  forward-first: no
```

1. On the "attacker" machine, install unbound and run `unbound -d -v -c unbound.conf`. 
2. On the machine running systemd-resolved, run `sudo systemd-resolve --set-dns 192.168.1.1:10053 --set-dnssec yes --interface enp0s3`. (Replace IP and interface name accordingly).
3. Run `resolvectl status` to confirm our DNS server is used and output contains `DNSSEC=yes/supported`.
4. Run `resolvectl query cloudflare.net` which should produce the correct reply and `Data is authenticated: yes`, confirming that the domain is DNSSEC-signed.
5. Run `resolvectl query www.cloudflare.net`:
```
$ resolvectl query www.cloudflare.net
www.cloudflare.net: 1.2.3.4                    -- link: enp0s3

-- Information acquired via protocol DNS in 25.6ms.
-- Data is authenticated: no; Data was acquired via local or encrypted transport: no
-- Data from: network
```
systemd-resolved accepts the bogus response. Running `dig @127.0.0.53 www.cloudflare.net` also shows the status is `NOERROR`. When resolved is configured as system resolver (e.g. through `/etc/resolv.conf`), `curl -v http://www.cloudflare.net` tries to connect to `1.2.3.4`, showing that applications are successfully tricked with the manipulated record.

Note that resolved does state the data is unauthenticated and also does not set the AD bit when queried on `127.0.0.53`.

Repeat the test but replace `typetransparent` with `static` in the unbound configuration. In this case, resolved rejects the response as expected:
```
$ resolvectl query www.cloudflare.net
www.cloudflare.net: resolve call failed: DNSSEC validation failed: no-signature
```
(Note that due to another bug https://github.com/systemd/systemd/issues/24827, resolved will still reply with the records on `127.0.0.53`, but it has `status: SERVFAIL`, so applications generally won't use this answer)

The difference is that `static` replaces all data for the `www.cloudflare.net` domain, whereas `typetransparent` only replaces the A record(s). Both strip the RRSIG records from affected queries.

My knowledge on DNS isn't great, but as far as I know no unsigned record should be allowed anywhere under the zone cloudflare.net, unless there was an insecure delegation (NS record without corresponding DS record).
However, it appears only signatures of some types of records actually matter to resolved, for example the SOA record. The RRSIG signatures on A/AAAA (and probably more) records can simply be stripped and resolved will accept the data without ensuring its validity.

### Additional program output to the terminal or log subsystem illustrating the issue

```sh
systemd-resolved debug logs, including all output after running `resolvectl query www.cloudflare.net` are posted in this gist, separate for the typetransparent and the static test case: https://gist.github.com/AgentOak/11c64ff8701c4cd853d6b99382a40704
```

**Comment by pemensik** [comment]
Yes, I confirm this is a serious bug. It allows downgrading to unsigned responses, which are accepted by resolved just fine. Anyone who just strips signatures can forge anything they want.

Reproduced that on ``systemd-253.5-6.fc39.x86_64``. I have used ``www.example.org``, which is signed also.

```
# resolvectl query -t a www.example.org
www.example.org IN A 127.0.0.1

# resolvectl
Global
         Protocols: LLMNR=resolve -mDNS -DNSOverTLS DNSSEC=yes/supported
  resolv.conf mode: stub
       DNS Servers: 127.0.0.1

Link 2 (host0)
    Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
         Protocols: +DefaultRoute LLMNR=resolve -mDNS -DNSOverTLS DNSSEC=yes/supported
Current DNS Server: 127.0.0.1
       DNS Servers: 127.0.0.1
# delv www.example.org
;; broken trust chain resolving 'www.example.org/A/IN': 127.0.0.53#53
;; resolution failed: broken trust chain
# delv @localhost www.example.org
;; insecurity proof failed resolving 'www.example.org/A/IN': 127.0.0.1#53
;; validating www.example.org/A: got insecure response; parent indicates it should be secure
;; insecurity proof failed resolving 'www.example.org/A/IN': ::1#53
;; resolution failed: insecurity proof failed
```

```
# drill -k /var/lib/unbound/root.key -D www.example.org
;; Number of trusted keys: 1
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 48044
;; flags: qr rd ra ; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; www.example.org.	IN	A

;; ANSWER SECTION:
www.example.org.	3600	IN	A	127.0.0.1

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 6 msec
;; EDNS: version 0; flags: do ; udp: 65494
;; SERVER: 127.0.0.53
;; WHEN: Wed Jul 12 12:55:30 2023
;; MSG SIZE  rcvd: 60
; www.example.org.	3600	IN	A	127.0.0.1
Bad data; RR for name and type not found or failed to verify, and denial of existence failed.


# drill -k /var/lib/unbound/root.key -D example.org
;; Number of trusted keys: 1
;; ->>HEADER<<- opcode: QUERY, rcode: NOERROR, id: 45411
;; flags: qr rd ra ad ; QUERY: 1, ANSWER: 2, AUTHORITY: 0, ADDITIONAL: 0 
;; QUESTION SECTION:
;; example.org.	IN	A

;; ANSWER SECTION:
example.org.	10501	IN	A	93.184.216.34
example.org.	10501	IN	RRSIG	A 13 2 86400 20230730225255 20230710002038 19897 example.org. 8cBc05Jk1h9pTPqfEvN4HPki3PPnoSbYUlz0INU6kQ+Tec+vYnbPcn6Nqy5fadmsLRzyyfTo6gJSky59hFONgw==

;; AUTHORITY SECTION:

;; ADDITIONAL SECTION:

;; Query time: 12 msec
;; EDNS: version 0; flags: do ; udp: 65494
;; SERVER: 127.0.0.53
;; WHEN: Wed Jul 12 12:56:01 2023
;; MSG SIZE  rcvd: 163
; example.org.	10501	IN	A	93.184.216.34
; No keys with the keytag and algorithm from the RRSIG found for id = 20326, owner = .
```

@keszybz @msekletar I would consider this high severity, unless DNSSEC can be not compiled in at all. This makes DNSSEC support just virtual and useless.

**Comment by pemensik** [comment]
Of course where signatures are not removed, it works fine. Of course ``ResolveUnicastSingleLabel=no`` is breaking validation from delv and others doing full verification from root.

```
# delv -t AAAA www.example.org @127.0.0.1
; fully validated
www.example.org.	43009	IN	AAAA	2606:2800:220:1:248:1893:25c8:1946
www.example.org.	43009	IN	RRSIG	AAAA 13 3 86400 20230730030327 20230709122038 19897 example.org. HPXAjF6ji9/BZ70QIFciHZFQ08ebcHFMC3wqmZDkCpLb6q1Zf+w11TcC qIFCscQlsNszkRhBjmmZwAZ67N+9Hw==
[root@rawhide ~]# delv -t AAAA www.example.org @127.0.0.53
;; broken trust chain resolving 'example.org/DS/IN': 127.0.0.53#53
;; broken trust chain resolving 'example.org/DNSKEY/IN': 127.0.0.53#53
;; broken trust chain resolving 'www.example.org/AAAA/IN': 127.0.0.53#53
;; resolution failed: broken trust chain

[root@rawhide ~]# delv -t AAAA www.example.org @127.0.0.54
; fully validated
www.example.org.	42853	IN	AAAA	2606:2800:220:1:248:1893:25c8:1946
www.example.org.	42853	IN	RRSIG	AAAA 13 3 86400 20230730030327 20230709122038 19897 example.org. HPXAjF6ji9/BZ70QIFciHZFQ08ebcHFMC3wqmZDkCpLb6q1Zf+w11TcC qIFCscQlsNszkRhBjmmZwAZ67N+9Hw==

[root@rawhide ~]# resolvectl 
Global
         Protocols: LLMNR=resolve -mDNS -DNSOverTLS DNSSEC=yes/supported
  resolv.conf mode: stub
       DNS Servers: 127.0.0.1

Link 2 (host0)
    Current Scopes: DNS LLMNR/IPv4 LLMNR/IPv6
         Protocols: +DefaultRoute LLMNR=resolve -mDNS -DNSOverTLS DNSSEC=yes/supported
Current DNS Server: 127.0.0.1
       DNS Servers: 127.0.0.1
```

**Comment by pemensik** [comment]
According to unbound logs, it did appropriate queries.

```
# journalctl -xeu unbound
Jul 12 12:36:08 rawhide unbound[1479]: [1479:0] info: start of service (unbound 1.17.1).
Jul 12 12:36:43 rawhide unbound[1479]: [1479:1] info: generate keytag query _ta-4f66. NULL IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:3] info: 127.0.0.1 www.example.org. A IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:0] info: 127.0.0.1 www.example.org. SOA IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:3] info: 127.0.0.1 example.org. DNSKEY IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:0] info: 127.0.0.1 www.example.org. DS IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:0] info: 127.0.0.1 example.org. DS IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:2] info: 127.0.0.1 example.org. SOA IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:3] info: 127.0.0.1 org. DNSKEY IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:3] info: 127.0.0.1 org. DS IN
Jul 12 13:16:29 rawhide unbound[1479]: [1479:1] info: 127.0.0.1 . DNSKEY IN

[root@rawhide ~]# delv +dnssec @localhost www.example.org ds
;; resolution failed: ncache nxrrset
; negative response, fully validated
; www.example.org.	4	IN	\-DS	;-$NXRRSET
; example.org. SOA ns.icann.org. noc.dns.icann.org. 2022091309 7200 3600 1209600 3600
; example.org. RRSIG SOA ...
; 2um39jhkjh9ji063sjoe2ojitictqjgq.example.org. NSEC3 1 0 5 CBC342D6EF20B86F SI4A4T6O7VP04OUCDB0Q9PI317K1UUC8 A TXT AAAA RRSIG
; 2um39jhkjh9ji063sjoe2ojitictqjgq.example.org. RRSIG NSEC3 ...

[root@rawhide ~]# delv +dnssec @localhost example.org dnskey
; fully validated
example.org.		2272	IN	DNSKEY	257 3 13 kRMX25/TJovOrsWq9Hv6QEFpzYsxItaOWPduFEwPz+5FM97SEHyCx+fc /XUN9gtktpXx45LAZpg/sFFEQH89og==  ; KSK; alg = ECDSAP256SHA256 ; key id = 22273
example.org.		2272	IN	DNSKEY	256 3 13 PznHCASFqSzeuHMYEdl6CauJtJUz6JbcyVYowRID9vN6Egsnn6NzhKbR RMReB6F0kXGy98LhiCn6TMG1W15nyA==  ; ZSK; alg = ECDSAP256SHA256 ; key id = 19897
example.org.		2272	IN	RRSIG	DNSKEY 13 2 3600 20230728174539 20230707221939 22273 example.org. uNx/5chJrQR1NqGNaEjrSN37rLafNCeZQPo7wlNbkGHIRhfV7QKpFVKM I0mJNPvmNIOFl0HEjMZQbaKu7LoYNg==
```

But it fails to deduce missing RRSIGs needs to be related to crytographic proof of DS record in its own zone or any of parent's. It is described in https://www.rfc-editor.org/rfc/rfc4035.html#section-5

>    The absence of DNSSEC data in a response MUST NOT by
>    itself be taken as an indication that no authentication information
>    exists.
> 
>    A resolver SHOULD expect authentication information from signed
>    zones.  A resolver SHOULD believe that a zone is signed if the
>    resolver has been configured with public key information for the
>    zone, or if the zone's parent is signed and the delegation from the
>    parent contains a DS RRset.

**Comment by pemensik** [comment]
It does not matter if I try that on zone (apex) name or any its member. If just *example.org* provides unsigned faked address, it will still be accepted by systemd-resolved, but correctly refused by other validating resolvers.

**Comment by evverx** [comment]
@pemensik just out of curiosity looking at the "rhel/downstream" label I wonder if RHEL is actually planning to include resolved in the foreseeable future?

FWIW this issue has been known since 2020: https://github.com/systemd/systemd/issues/15158

**Comment by pemensik** [comment]
Yes, people from systemd want systemd-resolved fully supported in RHEL 10. It is just *technology preview* in RHEL 8 and 9. They even proposed to have it enabled by default, just like in Fedora. I am not in favor of it, I would prefer to have *unbound* integrated with *NetworkManager* instead.

I wonder if ``dnssec`` tag actually means *ignore this, no one needs it* in this project.

**Comment by evverx** [comment]
> It is just technology preview in RHEL 8 and 9, the change of its status is under review. I don't like the way it is used in Fedora default installation

Got it. On the bright side it seems when it was turned into a technology preview feature it started getting tests upstream. It's far from what I'd expect but it's better than nothing I guess.

> I would choose unbound integration with NetworkManager myself

Looks like you aren't in favor of networkd either :-)

> I wonder if dnssec issue tag actually means ignore this, no one needs it in this project.

I don't know but given that my comment got downvoted it seems there was a miscommunication on my part. Sorry about that. I was just trying to figure out where resolved is headed. I agree it's not OK to ignore issues like this. The way it was handled on the systemd security mailing list isn't perfect either but I'm not subscribed to it and have no idea what's going on there. I generally report all systemd UAFs, double-frees, buffer overflows and stuff like that publicly.

**Comment by poettering** [comment]
The DNSSEC support in resolved is off by default for a reason, and it's not complete. It's very hard to get this work in a reasonable way given the state of DNSSEC servers in the wild. Before we can actually recommend people to use this we have to tweak this code a lot still. DNSSEC on the client is on the backburner, since it's such a massive amount of work to make it reasonably work with the end-user routers out there.

I am happy to review patches, but just bitching about an experimental feature all the time really doesn't help anyone. It just makes me lower my interest in dealing with these issues.

But yes, DNSSEC issues are not a high priority for us at the moment. Other DNS issues are more relevant. Also note that if we'd claim something was authenticated when it isn't this would be a much bigger issue. Here however, we simply don't authenticate, but say so clearly: "Data is authenticated: no"

**Comment by Winterhuman** [comment]
I was wondering why DNSSEC being considered experimental seemed odd to me, and figured out it was because none of the docs mention it being as such. I've opened a pull request to add a note: https://github.com/systemd/systemd/pull/28386

**Comment by AgentOak** [comment]
> FWIW this issue has been known since 2020: #15158

Huh, I completely missed this. I tried quite a few search terms before reporting this initially, but apparently didn't try "MITM", even though I ended up putting it in the title of this bug. Of course this 

> The DNSSEC support in resolved is off by default for a reason

```
Defaults to "allow-downgrade".
```

> Before we can actually recommend people to use this

```
It is recommended to set DNSSEC= to true on systems where it
is known that the DNS server supports DNSSEC correctly, and
where software or trust anchor updates happen regularly.
```

I understand that some distributions deviate from systemd's build-time defaults and the recommendation was written a long time ago, perhaps with the intent of "DNSSEC is nice to have if it works", but it appears there is a considerable disconnect between the user-facing [documentation](https://www.freedesktop.org/software/systemd/man/resolved.conf.html#DNSSEC=) and the state of things. As a regular user of systemd I was not aware of the experimental nature of resolved's DNSSEC support. By virtue of being part of systemd one would expect resolved to live up to the same standards of security and reliability unless indicated otherwise. 

Right now (and probably ever since its introduction) `DNSSEC=yes` does not give any additional security property over `DNSSEC=allow-downgrade`, whereas the documentation implies only `allow-downgrade` would be downgradable:

```
If set to "allow-downgrade" DNSSEC validation is attempted, but
if the server does not support DNSSEC properly, DNSSEC mode is
automatically disabled. Note that this mode makes DNSSEC
validation vulnerable to "downgrade" attacks.
```

> Also note that if we'd claim something was authenticated when it isn't this would be a much bigger issue

fwiw I agree, nothing really relies on DNSSEC security and the few tools that do have to check AD flag anyway. I just find it unfortunate that systemd claims security properties in resolved that have been known not to be true for over 3 years.

**Comment by peterthomassen** [comment]
> It's very hard to get this work in a reasonable way given the state of DNSSEC servers in the wild.

DNSSEC is well-specified; if things "in the wild" don't comply, it's OK to return an error. -- How does that interfere with getting it to work? Is that claim supported by any specific challenges?

This is not a cynical question. Other resolvers do get it to work, and I'm interested in identifying the problem(s) so I can help solve them.

**Comment by pemensik** [comment]
> The DNSSEC support in resolved is off by default for a reason, and it's not complete. It's very hard to get this work in a reasonable way given the state of DNSSEC servers in the wild. Before we can actually recommend people to use this we have to tweak this code a lot still. DNSSEC on the client is on the backburner, since it's such a massive amount of work to make it reasonably work with the end-user routers out there.

I would say the problem is systemd-resolved prevents usage of DNSSEC by local software, unless it has DNSSEC enabled itself. I have reported it on #19227. It was [you personally](https://github.com/systemd/systemd/issues/19227#issuecomment-814798010) refusing this idea. I would say the first mistake were to implement DNSSEC support from scratch, when there were libraries making that easier. ldns and unbound-libs might be a good examples. Even old issue #4621 did not request systemd should be doing flawless validation. It requested to pass DNSSEC records to tools and libraries understanding them (and handling them properly). It were closed, but the problem remains to be present on stub listener 127.0.0.53 until now.

> 
> I am happy to review patches, but just bitching about an experimental feature all the time really doesn't help anyone. It just makes me lower my interest in dealing with these issues.

Since when it is marked experimental? Can you point to *documentation* saying that?

From what I saw you interest in dealing with *dnssec* issues has been quite low since I got involved in DNS. I am not sure how different it would be if it gets worse. The problem is no-one else did look after them.

> But yes, DNSSEC issues are not a high priority for us at the moment. Other DNS issues are more relevant. Also note that if we'd claim something was authenticated when it isn't this would be a much bigger issue. Here however, we simply don't authenticate, but say so clearly: "Data is authenticated: no"

You are showing here the purpose of DNSSEC is not very clear to you. DNSSEC validation is not done to set AD flag when it were successful. It is done to prevent accepting bogus answers on any signed zone. Unsigned records have to be accepted only from unsigned zones, which have *DS* record missing or pointing to unsupported algorithm somewhere below. So if the ``resolvectl`` shows it does the validation, one expects signed domains like *fedoraproject.org* cannot accept fake response. Not authenticating just selected records, which can potential attacker can choose as he/she wishes, is a *security problem*. Saying just this one record from a signed zone were not authenticated is *not* acceptable, without negative trust anchor record at least. It then gives *false sense* of security, which might be *more* dangerous than just indicating clearly nothing is protected at all.

**Comment by pemensik** [comment]
> > The DNSSEC support in resolved is off by default for a reason
> 
> ```
> Defaults to "allow-downgrade".
> ```
It is overridden on both Fedora and Ubuntu builds to ``no``. On Fedora I am quite sure that was requested by someone from systemd team. https://fedoraproject.org/wiki/Changes/systemd-resolved#DNSSEC
> 
> Right now (and probably ever since its introduction) `DNSSEC=yes` does not give any additional security property over `DNSSEC=allow-downgrade`, whereas the documentation implies only `allow-downgrade` would be downgradable:

Correct. It makes it even worse, because you think you are protected, but would have to catch the insecure record only.

I think ``resolvectl query`` does not have ``--secure-only=yes``. Since there is ``resolvectl openpgp`` and ``resolvectl tlsa`` special options, one would expect they are safe to use. But have no way to use them securely only, especially with ``--raw`` output.

I have been searching for proofs why is DNSSEC so problematic since the Fedora change. I have found few references of really broken implementations, which responded with ``status: NXDOMAIN`` when query had DO=1, but ``status: NOERROR`` when query had DO=0. Such rare occurrences were clearly broken. Otherwise I have found just some forwarders respond without DNSSEC records even on DO=1 queries. Just like systemd-resolved with ``DNSSEC=no``. That is easy to detect and should be easily solved by ``allow-downgrade`` default.

I have not found strong evidence why DNSSEC needs to remain to ``no``. A bit tricky is detecting router, which does not understand DNSSEC, but is able to respond with signatures on the first uncached response. When responding from its own cache, it omits DNSSEC responses. That is more tricky to autodetect correctly. Some quite old dnsmasq behaved this way, it were fixed a long ago.

What I have found more often were various bugs in systemd-resolved implementation itself, which I have always reported. I were not very satisfied with the fixes usually, if there were any.

**Comment by pemensik** [comment]
> DNSSEC is well-specified; if things "in the wild" don't comply, it's OK to return an error. -- How does that interfere with getting it to work? Is that claim supported by any specific challenges?
> 
> This is not a cynical question. Other resolvers do get it to work, and I'm interested in identifying the problem(s) so I can help solve them.

I think other resolver vendors are confident enough to refuse doing workarounds to non-compliant DNS server answers. They require correct FORMERR from the server no understanding its query. Especially after [DNS flag days](https://www.dnsflagday.net/).

Another thing is mature resolvers require validation on/off boolean set by the administrator. They do not attempt to auto-detect its support on client side. I haven't seen alternative to ``allow-downgrade`` in any server-grade DNS forwarder.

The only alternative attempt to try such thing is [dnssec-trigger](https://www.nlnetlabs.nl/projects/dnssec-trigger/about/), I don't know others.

I think *systemd-resolved* tried to work even with *broken* servers and tried strange workarounds to prevent end-users complaining about regressions. But left other complaints ignored.

**Comment by bluca** [comment]
Given it seems necessary, let me restate the obvious: this is an open source project, and to the best of my knowledge none of the maintainers or core developers are paid to work on dnssec. So, any time spent excessively arguing on the bug tracker on an issue is already opened and triaged, is time that is pretty much wasted, as it will not cause a bug fix to conjure itself out of thin air. Instead, for those who care about such use cases, it would be much, much better to spend the time to work on a PR to fix such issues, and then we can review that. Thanks.

**Comment by pemensik** [comment]
I am willing to review PR, but I am not creating it. I am not paid for working on *systemd* but on *dnsmasq*, *unbound* and *bind9*. All of them have no such issue, but for a reason hard to understand *none* of them are the default choice. They are often broken by *systemd-resolved* deployment, hindering my work. Thanks for understanding.

**Comment by bluca** [comment]
Well, it's not really hard to understand at all: those are affected by much more important, more relevant and more annoying issues that users actually care about, as what users want from their internet connection is for it to _work_ and fetch those cat pictures and fetch them fast, everything else is pretty much secondary or tertiary to that. Cat pictures first, protocol details last. But I digress. Once again, writing more and more and more comments on the same already triaged issue is not going to make a fix appear out of nowhere, somebody needs to volunteer or get paid to work on it.

**Comment by pemensik** [comment]
Note: *systemd-resolved* is *preinstalled* also on Fedora *Server* and maybe other distributions targeting server deployments too. The same limitation applies there and is more severe. I haven't seen any declaration it is not suitable for such use. Please be so kind and point me to some.

If anyone needs clarification on any DNS protocol parts or RFC compliance, I would recommend visiting [OARC Mattermost](https://chat.dns-oarc.net/). It is place full of DNS professionals often willing to help.

I would recommend trying ``delv +vtrace @127.0.0.54 example.org``. It explains a lot of done steps and may provide terms to google for. Compare results with ``delv +vtrace @127.0.0.53 example.org``.

**Comment by bluca** [comment]
> Once again, writing more and more and more comments on the same already triaged issue is not going to make a fix appear out of nowhere, somebody needs to volunteer or get paid to work on it.

**Comment by carnil** [comment]
Doing some CVE triage in another downstream distribution, I noticed https://bugzilla.redhat.com/show_bug.cgi?id=2222672 which has CVE-2023-7008 and seem to relate to this problem.

**Comment by AgentOak** [comment]
[Red Hat CVE-2023-7008 entry](https://access.redhat.com/security/cve/CVE-2023-7008) currently shows

> Public on December 8, 2022

However it should be noted that this issue was publicly known since March 18, 2020, as it was previously reported here: https://github.com/systemd/systemd/issues/15158

**Comment by bluca** [comment]
This is extremely confusing - who requested a CVE and why? Why weren't we consulted ahead of time? Also, the classification is bogus: this requires local root privileges or user interaction to enable in the first place, given it's disabled by default, and yet that link says "Privileges Required: none" and "User Interaction: none"

**Comment by flokli** [comment]
> Also, the classification is bogus: this requires local root privileges or user interaction to enable in the first place, given it's disabled by default, and yet that link says "Privileges Required: none" and "User Interaction: none"

As per https://www.freedesktop.org/software/systemd/man/latest/resolved.conf.html#DNSSEC=, it doesn't default to "false", but "allow-downgrade", and there's nothing in the docs stating the feature is known to be broken:

> If set to true, all DNS lookups are DNSSEC-validated locally (excluding LLMNR and Multicast DNS). If the response to a lookup request is detected to be invalid a lookup failure is returned to applications. Note that this mode requires a DNS server that supports DNSSEC. If the DNS server does not properly support DNSSEC all validations will fail.

If people rely on this to work, set the option to true, they expect their DNS lookups to not be MITM'able.

It turns out, they unexpectedly are, this can have security/privacy implications, so it's fine to request a CVE, so that issue (and when it's fixed) can be tracked with CVE tools.

**Comment by bluca** [comment]
> > Also, the classification is bogus: this requires local root privileges or user interaction to enable in the first place, given it's disabled by default, and yet that link says "Privileges Required: none" and "User Interaction: none"
> 
> As per https://www.freedesktop.org/software/systemd/man/latest/resolved.conf.html#DNSSEC=, it doesn't default to "false", but "allow-downgrade", and there's nothing in the docs stating the feature is known to be broken:

Doesn't change anything, this is a real issue only if it's enforced, ie =yes, with implicit downgrades it's pretty much cosmetic

> > If set to true, all DNS lookups are DNSSEC-validated locally (excluding LLMNR and Multicast DNS). If the response to a lookup request is detected to be invalid a lookup failure is returned to applications. Note that this mode requires a DNS server that supports DNSSEC. If the DNS server does not properly support DNSSEC all validations will fail.
> 
> If people rely on this to work, set the option to true, they expect their DNS lookups to not be MITM'able.
> 
> It turns out, they unexpectedly are, this can have security/privacy implications, so it's fine to request a CVE, so that issue (and when it's fixed) can be tracked with CVE tools.

The problem is the process: as far as I'm aware nobody contacted us and said "hey we want to request a CVE for this, are you fine with it?", and that's very much not ok. So once again, who requested it?

**Comment by AgentOak** [comment]
> Why weren't we consulted ahead of time?

> Please note that I have repeatedly tried to report this issue to the systemd-security mailing list. I sent a mail with the details and reproduction steps to the systemd-security address on Oct 6th, Oct 14th and Oct 22nd. I first received a reply on Oct 24th from Lennart Poettering, asking for some basic system information, which I provided on the same day. Then I received no further reply, so I sent the report one more time to systemd-security on Nov 7th. The next day Poettering replied to my previous mail, asking for a debug log, but not directly acknowledging the issue. I sent debug logs on Nov 10th and haven't heard back from anyone since then.

It did not appear that systemd had an interest in this matter. This is further underlined by the fact that this has been a known security vulnerability in systemd for nearly three years.

> this requires local root privileges or user interaction to enable in the first place

I don't see how that makes sense. CVEs also apply to non-default configuration. Why would an attacker enable DNSSEC to then attack DNSSEC validation? The issue is about attacking a machine running with `DNSSEC=true`.

**Comment by peterthomassen** [comment]
> The problem is the process: as far as I'm aware nobody contacted us and said "hey we want to request a CVE for this, are you fine with it?", and that's very much not ok.

I guess this is your opinion.

> So once again, who requested it?

Perhaps this is not the right forum to figure that out.

**Comment by bluca** [comment]
> > The problem is the process: as far as I'm aware nobody contacted us and said "hey we want to request a CVE for this, are you fine with it?", and that's very much not ok.
> 
> I guess this is your opinion.

And?

> > So once again, who requested it?
> 
> Perhaps this is not the right forum to figure that out.

Which one would that be then?

**Comment by pemensik** [comment]
> The problem is the process: as far as I'm aware nobody contacted us and said "hey we want to request a CVE for this, are you fine with it?", and that's very much not ok. So once again, who requested it?

I did, I requested a review by Red Hat security response team.  I am pretty sure the report were forwarded to some systemd maintainers in Red Hat also. Unlike you guys waiving this as *not* important, I think this is *security* related issue and people using DNSSEC in systemd-resolved expect they are protected (and documenting the opposite were refused).

You guys were reported this issue, which has severity implications and ignored that for *years*. Reporting the same thing again to *same* people did not seem work better. Especially after discussion here.

**Comment by msekletar** [comment]
> I am pretty sure the report were forwarded to some systemd maintainers in Red Hat also.

Not sure about other RH maintainers but I haven't received any notification that CVE was requested for this issue.

**Comment by pemensik** [comment]
> The problem is the process: as far as I'm aware nobody contacted us and said "hey we want to request a CVE for this, are you fine with it?", and that's very much not ok. So once again, who requested it?

Usual workflow with vulnerability report received in other projects I work on is to acknowledge the vulnerability and prepare fix ASAP. You as upstream do not need a CVE assigned to prepare a fix present at least in development versions. If reported publicly, it should be done faster if possible.



> > I am pretty sure the report were forwarded to some systemd maintainers in Red Hat also.
> 
> Not sure about other RH maintainers but I haven't received any notification that CVE was requested for this issue.

https://bugzilla.redhat.com/show_bug.cgi?id=2222672#c3 proves @lnykryn reacted on it. Sorry if it missed you. I admit the CVE bugzilla were missing direct link to this issue until today in comment 5.

**Comment by bluca** [comment]
> > The problem is the process: as far as I'm aware nobody contacted us and said "hey we want to request a CVE for this, are you fine with it?", and that's very much not ok. So once again, who requested it?
> 
> I did, I requested a review by Red Hat security response team. I am pretty sure the report were forwarded to some systemd maintainers in Red Hat also. Unlike you guys waiving this as _not_ important, I think this is _security_ related issue and people using DNSSEC in systemd-resolved expect they are protected (and documenting the opposite were refused).
> 
> You guys were reported this issue, which has severity implications and ignored that for _years_. Reporting the same thing again to _same_ people did not seem work better. Especially after discussion here.

Yeah no, that is not how any of this works. The fact that the issue is not a priority for anybody here doesn't mean you can go behind our backs and get CVEs assigned to a project that you are not part of, without notifying anybody before or after, and ignoring well-established processes to handle situations like this. I found out because the Debian security team asked me, this is just absurd. It's all the more absurd because a coordinated CVE, with the appropriate description and score, would have been just fine.

I consider this is a blatant abuse of Redhat's status as CNA. In order to prove that Redhat can be trusted as a CNA, what you need to do right now, given you are the submitter, is to get the advisory updated so that it correctly and prominently states that local root access is needed to enable the affected feature, which is disabled by default, and that in case this hasn't happened users are unaffected. The score needs to be downgraded significantly as a result.

**Comment by msekletar** [comment]
> https://bugzilla.redhat.com/show_bug.cgi?id=2222672#c3 proves @lnykryn reacted on it. Sorry if it missed you. I admit the CVE bugzilla were missing direct link to this issue until today in comment 5.

At the time when @lnykryn was reacting to the bug it had no CVE assigned, as bug got updated just yesterday. I was aware of the issue and that RH ProdSec team is tracking it (due to FedRamp) what I was not aware (and I bet neither @lnykryn) is that CVE process has been started.

**Comment by pemensik** [comment]
That proves you were aware it was vulnerability bug and were given enough time to assign it own CVE by own process, whatever it is. But someone else did that, I wonder why?

**Comment by msekletar** [comment]
> But someone else did that, I wonder why?

The problem isn't that someone else assigned CVE, problem is they didn't talk to anyone from the systemd team, neither upstream nor internally at Red Hat.

**Comment by pemensik** [comment]
I am afraid this issue is still not fixed. Made my example for reproduction on: https://github.com/systemd/systemd/pull/30549#issuecomment-1868013733. It still allows forging any record into signed zone, but now maybe with exception of SOA query. That is not enough, sorry.

**Comment by pemensik** [comment]
> > But someone else did that, I wonder why?
> 
> The problem isn't that someone else assigned CVE, problem is they didn't talk to anyone from the systemd team, neither upstream nor internally at Red Hat.

The problem is Security team haven't found any progress in fixing this issue, so they are even considering preparing in-house fix.

**Comment by bluca** [comment]
That's not really important at all right now - I see the advisory has still not been updated, what's the ETA on that?

**Comment by pemensik** [comment]
> That's not really important at all right now - I see the advisory has still not been updated, what's the ETA on that?

I am not sure what update you would like. In ``DNSSEC=no`` it is possible to always spoof records, but only with ``DNSSEC=yes`` it can be vulnerability. I hope you refer to [CVE-2023-7008](https://access.redhat.com/security/cve/CVE-2023-7008) page, [CVE](https://www.cve.org/CVERecord?id=CVE-2023-7008) entries on cve.org usually take quite long to update. I have added @msekletar to my ticket with security team, he should be able to negotiate anything important.

**Comment by bluca** [comment]
https://github.com/systemd/systemd/issues/25676#issuecomment-1867914879

> I consider this is a blatant abuse of Redhat's status as CNA. In order to prove that Redhat can be trusted as a CNA, what you need to do right now, given you are the submitter, is to get the advisory updated so that it correctly and prominently states that local root access is needed to enable the affected feature, which is disabled by default, and that in case this hasn't happened users are unaffected. The score needs to be downgraded significantly as a result.

Have you requested such changes and have you been given an ETA for when https://access.redhat.com/security/cve/CVE-2023-7008 is going to be updated to reflect reality?

**Comment by pemensik** [comment]
I am not a member of Security response team. Michal knows how to reach them and how to make requests. All I can do is comments in https://bugzilla.redhat.com/show_bug.cgi?id=2222672, which I believe describes pretty well current state and the vulnerability. I believe upstream and downstream vendor CVSS score might be different. I have never tried to generate it and think security people are better suited for defining it.

**Comment by bluca** [comment]
You said here https://github.com/systemd/systemd/issues/25676#issuecomment-1867901065 that you are the one that filed it, which means you can also request to update it to fix the  as needed. This is the third time now you have been explicitly asked to remediate to this mistake, are you going to do that or not?

**Comment by pemensik** [comment]
What I did was sending request to our *secalert*. From there onward it is their job and responsibility to evaluate. I have no edit rights to any CVE including for my own components, which systemd is not. I admit the timing is quite bad now, but all I can do is writing comments.

This attack affects only instances configured by root to set ``DNSSEC=yes``, that is correct. Not explicitly But anyone able to send packets accepted by resolved, meaning usually on-path attacker or attacker able to guess port and query id. It can be misused by external entity over network, without having local root rights. It is unclear what correction you are demanding.

**Comment by bluca** [comment]
> What I did was sending request to our _secalert_. From there onward it is their job and responsibility to evaluate. I have no edit rights to any CVE including for my own components, which systemd is not. I admit the timing is quite bad now, but all I can do is writing comments.

The ask was not to update it directly, as you can see in the three comments above - the ask was to request to update it as indicated, to fix the obvious mistakes in the advisory that have already been pointed out multiple times now. Given you could request it in the first place, it's obvious you can also request it to be fixed as needed. This is the 4th time now. Are you going to do that or do we have to start assuming this was not an honest mistake after all?

**Comment by pemensik** [comment]
Oh wait, maybe just systemd-resolved Fedora package does not properly restart itself. After upgrade *and* manual ``systemctl restart systemd-resolved`` it seems to be indeed fixed correctly.

```
# resolvectl query -t a example.net 
example.net: resolve call failed: DNSSEC validation failed: no-signature
# resolvectl query -t txt example.net 
example.net: resolve call failed: DNSSEC validation failed: no-signature
```

Still with the same package I tried before. It is unexpected dig returns both SERVFAIL *and* invalid answer. No good DNS implementations do that, but I am not sure it is strictly defined. I think I have filled some issue for it.

```
# dig example.net a

; <<>> DiG 9.18.20 <<>> example.net a
;; global options: +cmd
;; Got answer:
;; ->>HEADER<<- opcode: QUERY, status: SERVFAIL, id: 27889
;; flags: qr rd ra; QUERY: 1, ANSWER: 1, AUTHORITY: 0, ADDITIONAL: 1

;; OPT PSEUDOSECTION:
; EDNS: version: 0, flags:; udp: 65494
;; QUESTION SECTION:
;example.net.			IN	A

;; ANSWER SECTION:
example.net.		0	IN	A	127.0.0.255

;; Query time: 0 msec
;; SERVER: 127.0.0.53#53(127.0.0.53) (UDP)
;; WHEN: Fri Dec 22 22:59:10 CET 2023
;; MSG SIZE  rcvd: 56

# dig +short example.net
127.0.0.255
# dig +short example.net txt
"v=spf1 +all"
```

Which may make some clients accept even this invalid response, if they do not check status properly. It looks fixed, mostly.

**Comment by bluca** [comment]
> > What I did was sending request to our _secalert_. From there onward it is their job and responsibility to evaluate. I have no edit rights to any CVE including for my own components, which systemd is not. I admit the timing is quite bad now, but all I can do is writing comments.
> 
> The ask was not to update it directly, as you can see in the three comments above - the ask was to request to update it as indicated, to fix the obvious mistakes in the advisory that have already been pointed out multiple times now. Given you could request it in the first place, it's obvious you can also request it to be fixed as needed. This is the 4th time now. Are you going to do that or do we have to start assuming this was not an honest mistake after all?

Still waiting for an answer

**Comment by pemensik** [comment]
@bluca I do not think there is anything important to update. Ask @msekletar to request any changes, as he may understand your motivation better. I see nothing important to change there. I would not agree to any attempts making it look less important than it is already.

**Comment by pemensik** [comment]
Okay, the fix is sufficient after all, false alarm. Likely problem to be solved in Fedora packaging, but with restart it is fixed. Strange answer even with servfail has own issue #24827, therefore this can be truly closed.

**Comment by bluca** [comment]
This is not a debate or an exchange of opinions. You have made a series of grave mistakes, and 5 times you were offered the chance to remediate, and 5 times instead you chose to double down. You leave me no other option than considering your actions intentionally malicious and fraudulent, which means I'll have to act accordingly.

**Comment by pemensik** [comment]
My only mistake is I still try to communicate with and improve systemd. You have marked spam and off-topic comments, which were on-topic and to the point, just disagreeing with you. I am wasting my time here just because people deserve working name resolution and unfortunately *systemd-resolved* became default on some systems I work on. I am not sure what is your role in systemd @bluca, but I dislike your demands and accusations.

Merry christmas or Happy holidays. I will be on Fosdem in DNS room, if anyone wants to talk.

**Comment by bluca** [comment]
Given to date your code contributions are exactly zero, as anybody can verify from the git log, it seems a bit far fetched to claim you have improved anything. Instead, you have abused your position and internal processes to stir unnecessary troubles and publish intentionally misleading information, and it's pretty transparent why you are doing that. And on a week where most people here (myself included) are already on holiday, on top of that, which at this point one can reasonably assume to be intentional too. This is the exact opposite of "improving". It didn't have to turn out like this, and you had plenty of chances, but you chose this path again, and again, and again. Enough.

**Comment by peterthomassen** [comment]
For the record, I violently disagree that @voidpointertonull's comments qualify as spam or any other type of message that should be sanctioned. The project's communications in the issue are a shame.

**Comment by evverx** [comment]
@keszybz @poettering as far as I can tell Petr was removed from the organization and I really wonder what sort of process with made-up rules involving pressure and intimidation was followed here? I'd assume it was https://github.com/systemd/systemd/blob/main/docs/CODE_OF_CONDUCT.md but it was violated here quite a few times.

@keszybz since you are on the team that is supposed to enforce those things I expect the following clause
> All complaints will be reviewed and investigated and will result in a response that is deemed necessary and appropriate to the circumstances

to kick in here. Thank you.

**Comment by yuwata** [comment]
> For the record, I violently disagree that @voidpointertonull's comments qualify as spam or any other type of message that should be sanctioned. The project's communications in the issue are a shame.

I unhide the comments (and yours). These are not related to what Luca asked, but not off topic for this thread.

**Comment by yuwata** [comment]
> Petr was removed from the organization

@evverx What does it mean? At least, he is not blocked by this project. Luca just locked this thread.

**Comment by evverx** [comment]
> What does it mean?

@yuwata Petr was part of the systemd organization on GitHub and could comment on locked threads for example.

**Comment by yuwata** [comment]
I did not know that he was in systemd organization on GitHub, and I have no idea how to check that.

But anyway, I do not think here is not a good place to discuss about the removal here. That's completely off-topic here. Please use our internal chat or so.

**Comment by bluca** [comment]
> I unhide the comments (and yours). These are not related to what Luca asked, but not off topic for this thread.

These comments were completely off topic as the bug at hand was already solved, and clearly written to needlessly raise tension while adding nothing of value - this is an issue tracker, not a social media platform for the peanut gallery. So please leave them as such, thank you

**Comment by bluca** [comment]
> I did not know that he was in systemd organization on GitHub, and I have no idea how to check that.
> 
> But anyway, I do not think here is not a good place to discuss about the removal. That's completely off-topic here. Please use our internal chat or so.

That would have been better yes, but given this was raised publicly: Petr abused the internal cve process at Redhat, going behind the back of both the upstream and downstream teams, at a time when he knew most of the people involved were on holiday and thus unable to react without sacrificing personal time, providing intentionally misleading information to the security team, in order to personally benefit from the the issue to promote his own alternative projects on social media. When asked to rectify, 5 times, he refused and doubled down, 5 times. It's also worth noting that he did absolutely nothing to help here - the actual bug was opened by somebody else, and the solution was provided by Michal, and the backports and releases by myself. His only involvement here, beside the aforementioned abuse, was to raise tension and flame.

This plainly hostile and fraudulent behavior is obviously incompatible with membership of the org which, as a reminder, is not an inalienable right but granted at the pleasure of the org owners, who have elected to remove it. And this is just the beginning, the gravity of his actions are such that there surely will be more consequences in due time.

In the meanwhile, I want to thank @msekletar both for providing a fix and for raising the issue of the missing information from the cve with the internal team, which has now been updated and I am satisfied with how it's presented now. It was such a simple change, the submitter could have easily done so when asked the first time, and all of this could have been avoided.

**Comment by bluca** [comment]
Finally, the backports are going through CI, and stable releases for all affected versions should be ready by tomorrow.

**Comment by evverx** [comment]
> that any installation of systemd on any distribution was automatically affected

I'm extremely confused. It has been mentioned a lot of times here and on bugzilla explicitly that if affects things with `DNSSEC=yes` only and it's off by default. I can't see where anyone has denied that.

> has now been updated and I am satisfied with how it's presented now

As far as I can see https://access.redhat.com/security/cve/CVE-2023-7008 says now "Note that DNSSEC= option is by default set to 'no' and changing the value requires root privileges." but as far as I can tell the whole argument was whether the vector should be "AV:N/AC:H/PR:N/UI:N/S:U/C:N/I:H/A:N"

Is it going to be changed to "AV:L/AC:H/PR:H/UI:R/S:U/C:N/I:H/A:N" with local access, root privileges and user interaction?

**Comment by bluca** [comment]
> Finally, the backports are going through CI, and stable releases for all affected versions should be ready by tomorrow.

Ok the following backported commits with associated stable tags have been pushed:

https://github.com/systemd/systemd-stable/commit/6da5ca9dd69c0e3340d4439413718ad4963252de (v255.2)
https://github.com/systemd/systemd-stable/commit/029272750fe451aeaac87a8c783cfb067f001e16 (v254.8)
https://github.com/systemd/systemd-stable/commit/5c149c77cbf7b3743fa65ce7dc9d2b5a58351968 (v253.15)
https://github.com/systemd/systemd-stable/commit/bb78da7f955c0102047319c55fff9d853ab7c87a (v252.21)
https://github.com/systemd/systemd-stable/commit/f58fc88678b893162f2d6d4b2db094e7b1646386 (v251.20)
https://github.com/systemd/systemd-stable/commit/4ada1290584745ab6643eece9e1756a8c0e079ca (v250.14)
https://github.com/systemd/systemd-stable/commit/c8578cef7f0f1e8cb8193c29e5e77daf4e3a1c9f (v249.17)
https://github.com/systemd/systemd-stable/commit/3a409b210396c6a0bef621349f4caa3a865940f2 (v248.13)

There are reports that the issue appears as far back as v241, but before v248 the fix doesn't cleanly apply anymore as the branches are too old, as v247 was released 3 years ago and received the last non-cve-related maintenance release 2 years ago, and the rest is even older. Providing backports for the past 8 releases is enough work already, especially given it's Christmas, and more than it's usually done, for anything older the downstream maintainers of distributions that use any such old branches can take care of it, if they deem it worth it.

It was truly lovely to be forced to work on Christmas eve and the day before while on holiday instead of spending time with my family, but everything is now as sorted as it can be. The only thing left to do, once we are back from the holidays, is having a serious conversation about flaws in the internal CVE process at Redhat that allowed an individual to deliberately abuse it in such a way to inflict maximum pain, in order to extract personal profit for himself.

**Comment by evverx** [comment]
> The only thing left to do ... is having a serious conversation about flaws in the internal CVE process

I'm sorry but I'm not sure why all the blame is being shifted here far away from systemd. As far as systemd is concerned I think it would be much more productive to try to figure out why it happened the way it happened and what could have prevented it.

This whole thing was reported back in 2020. It could have been handled there properly. (It could have been documented that this is experimental and shouldn't be relied on but to judge from a PR that attempted to do exactly that it isn't desirable to clearly communicate the status of certain parts of systemd. That PR was pressured into submission and I can't seem to say that the participants there were happy with the communication style either).

Then this thing was reported to the "security" mailing list properly and got ghosted. It could have been handled there properly too.

Then this thing was reported here. The experimental status was acknowledged but it isn't exactly what is usually expected in response to issues reported on security mailing lists.

Then it escalated somehow and was blown out of proportion with threats and so on like it's 1998 again. I'm sorry but I don't think It looks like a mature process at all.

**Comment by mbiebl** [comment]
Someone is trying to sabotage systemd-resolved in Fedora it seems
https://www.spinics.net/lists/fedora-devel/msg317522.html
